### PR TITLE
Add Erased AST

### PR DESCRIPTION
**Pull request by JonathanStarup**

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+
+  sealed trait Expression[+T <: JType] {
+    def loc: SourceLocation
+  }
+
+  sealed trait JType
+
+  object JType {
+
+    sealed trait JByte8 extends JType
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala:336]: Lets just the Flix names- they seam more regular:

JByte8 -> PrimInt8
JShort16 -> PrimInt16

And then we need to add:

BoxedInt8
BoxedInt16

 and so forth

File: main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+
+    sealed trait JLong64 extends JType
+
+    sealed trait JChar extends JType
+
+    sealed trait JFloat32 extends JType
+
+    sealed trait JDouble64 extends JType
+
+    sealed trait JBool extends JType
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala:350]: There are no Booleans. There are PrimInt32 to branch on.

File: main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+
+    sealed trait JObject extends JType
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala:352]: In some sense, we can start where we only have the above types + the boxed variants.

Then we can erase everything to JObject (that is not e.g. an integer).

Then, over time, we can start using precise types like e.g. JRef below.

Does that make sense?

File: main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+//  case class Property(law: Symbol.DefnSym, defn: Symbol.DefnSym, exp: ErasedAst.Expression[_]) {
+//    def loc: SourceLocation = defn.loc
+//  }
+//
+//  case class LatticeOps(tpe: MonoType, bot: Symbol.DefnSym, top: Symbol.DefnSym, equ: Symbol.DefnSym, leq: Symbol.DefnSym, lub: Symbol.DefnSym, glb: Symbol.DefnSym, loc: SourceLocation)
+
+  object Expression {
+
+    case class Cast[T <: JType, S <: JType](exp: ErasedAst.Expression[T], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[S]
+
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala:50]: We will also need ops like:

`BoxInt8` and `UnboxInt16` and so forth.

File: main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+                  properties: List[ErasedAst.Property],
+                  specialOps: Map[SpecialOperator, Map[MonoType, Symbol.DefnSym]],
+                  reachable: Set[Symbol.DefnSym],
+                  sources: Map[Source, SourceLocation])
+
+  case class Def(ann: Ast.Annotations, mod: Ast.Modifiers, sym: Symbol.DefnSym, formals: List[ErasedAst.FormalParam], exp: ErasedAst.Expression[JType], tpe: MonoType, loc: SourceLocation) {
+    var method: Method = _
+  }
+
+  case class Enum(mod: Ast.Modifiers, sym: Symbol.EnumSym, cases: Map[Name.Tag, ErasedAst.Case], tpeDeprecated: MonoType, loc: SourceLocation)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala:40]: Remove `tpeDeprecated`.

File: main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+
+  case class Property(law: Symbol.DefnSym, defn: Symbol.DefnSym, exp: ErasedAst.Expression[JUnknown[Nothing]]) {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala:42]: JUnknown -> JBool

File: main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala [hunk 7]
```diff
@@ -0,0 +1,364 @@
...
+    case class BoxInt64(exp: ErasedAst.Expression[PrimInt64], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[BoxedInt64]
+    case class UnboxInt64(exp: ErasedAst.Expression[BoxedInt64], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[PrimInt64]
+    case class BoxChar(exp: ErasedAst.Expression[PrimChar], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[BoxedChar]
+    case class UnboxChar(exp: ErasedAst.Expression[BoxedChar], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[PrimChar]
+    case class BoxFloat32(exp: ErasedAst.Expression[PrimFloat32], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[BoxedFloat32]
+    case class UnboxFloat32(exp: ErasedAst.Expression[BoxedFloat32], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[PrimFloat32]
+    case class BoxFloat64(exp: ErasedAst.Expression[PrimFloat64], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[BoxedFloat64]
+    case class UnboxFloat64(exp: ErasedAst.Expression[BoxedFloat64], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[PrimFloat64]
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala:63]: Group:
```
case class BoxInt8
case class BoxInt16
```

and move to bottom of the `Expression` object.

File: main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala [hunk 8]
```diff
@@ -0,0 +1,364 @@
...
+
+    case class Unit(loc: SourceLocation) extends ErasedAst.Expression[JObject] {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala:65]: JObject -> JUnit

File: main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala [hunk 9]
```diff
@@ -0,0 +1,364 @@
...
+    }
+
+    case class Str(lit: java.lang.String, loc: SourceLocation) extends ErasedAst.Expression[JObject] {
+      final val tpe = MonoType.Str
+    }
+
+    case class Var[T <: JType](sym: Symbol.VarSym, tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[T]
+
+    // TODO: Get rid of the fnMonoType here.
+    case class Closure(sym: Symbol.DefnSym, freeVars: List[FreeVar], fnMonoType: MonoType, tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[JObject]
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala:118]: fnMonoType  -> remove

File: main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala [hunk 10]
```diff
@@ -0,0 +1,364 @@
...
+
+    case class ApplyClo(exp: ErasedAst.Expression[JObject], args: List[ErasedAst.Expression[JType]], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[JType]
+
+    case class ApplyDef(sym: Symbol.DefnSym, args: List[ErasedAst.Expression[JType]], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[JType]
+
+    case class ApplyCloTail(exp: ErasedAst.Expression[JObject], args: List[ErasedAst.Expression[JType]], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[JType]
+
+    case class ApplyDefTail(sym: Symbol.DefnSym, args: List[ErasedAst.Expression[JType]], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[JType]
+
+    case class ApplySelfTail(sym: Symbol.DefnSym, formals: List[ErasedAst.FormalParam], actuals: List[ErasedAst.Expression[JType]], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[JType]
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala:128]: Alle applys: [T <: JType]

File: main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala [hunk 11]
```diff
@@ -0,0 +1,364 @@
...
+
+    // TODO: maybe make multiple classes for different exp types
+    case class Unary(sop: SemanticOperator, op: UnaryOperator, exp: ErasedAst.Expression[JType], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[JType]
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala:131]: [T <: JType] for result

File: main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala [hunk 12]
```diff
@@ -0,0 +1,364 @@
...
+
+    case class Binary(sop: SemanticOperator, op: BinaryOperator, exp1: ErasedAst.Expression[JType], exp2: ErasedAst.Expression[JType], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[JType]
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala:133]: [T <: JType] for result

File: main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala [hunk 13]
```diff
@@ -0,0 +1,364 @@
...
+
+    case class IfThenElse[T <: JType](exp1: ErasedAst.Expression[PrimInt32], exp2: ErasedAst.Expression[T], exp3: ErasedAst.Expression[T], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[T]
+
+    case class Branch[T <: JType](exp: ErasedAst.Expression[JUnknown[Nothing]], branches: Map[Symbol.LabelSym, ErasedAst.Expression[T]], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[T]
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala:137]: JUnknown[Nothing] -> T

File: main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala [hunk 14]
```diff
@@ -0,0 +1,364 @@
...
+
+    case class JumpTo(sym: Symbol.LabelSym, tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[JUnknown[Nothing]]
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala:139]: [T <: JType] for result

File: main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala [hunk 15]
```diff
@@ -0,0 +1,364 @@
...
+
+    case class Let[T <: JType](sym: Symbol.VarSym, exp1: ErasedAst.Expression[JType], exp2: ErasedAst.Expression[T], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[T]
+
+    case class Is(sym: Symbol.EnumSym, tag: Name.Tag, exp: ErasedAst.Expression[JObject], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[PrimInt32]
+
+    case class Tag(sym: Symbol.EnumSym, tag: Name.Tag, exp: ErasedAst.Expression[JType], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[JObject]
+
+    case class Untag(sym: Symbol.EnumSym, tag: Name.Tag, exp: ErasedAst.Expression[JObject], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[JType]
+
+    case class Index(base: ErasedAst.Expression[JObject], offset: scala.Int, tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[JObject]
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala:149]: Result -> JType

File: main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala [hunk 16]
```diff
@@ -0,0 +1,364 @@
...
+
+    case class ArrayLength[T <: JType](base: ErasedAst.Expression[JArray[T]], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[PrimInt32]
+
+    case class ArraySlice[T <: JType](base: ErasedAst.Expression[JArray[T]], beginIndex: ErasedAst.Expression[PrimInt32], endIndex: ErasedAst.Expression[PrimInt32], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[JArray[T]]
+
+    case class Ref[T <: JType](exp: ErasedAst.Expression[T], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[JRef[T]]
+
+    case class Deref[T <: JType](exp: ErasedAst.Expression[JRef[T]], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[T]
+
+    case class Assign(exp1: ErasedAst.Expression[JType], exp2: ErasedAst.Expression[JType], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[JUnit]
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala:177]: `exp1` : Ref[T]
`exp2`: T

File: main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala [hunk 17]
```diff
@@ -0,0 +1,364 @@
...
+
+    case class GetField(field: Field, exp: ErasedAst.Expression[JObject], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[JType]
+
+    case class PutField(field: Field, exp1: ErasedAst.Expression[JObject], exp2: ErasedAst.Expression[JType], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[JUnit]
+
+    case class GetStaticField(field: Field, tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[JType]
+
+    case class PutStaticField(field: Field, exp: ErasedAst.Expression[JType], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[JUnit]
+
+    case class NewChannel[T <: JType](exp: ErasedAst.Expression[T], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[JChan[T]]
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala:201]: exp: PrimInt32

File: main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala [hunk 18]
```diff
@@ -0,0 +1,364 @@
...
+
+    case class SelectChannel[T <: JType](rules: List[ErasedAst.SelectChannelRule[T]], default: Option[ErasedAst.Expression[T]], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[T]
+
+    case class Spawn(exp: ErasedAst.Expression[JType], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[JUnit]
+
+    case class Lazy[T <: JType](exp: ErasedAst.Expression[T], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[JLazy[T]]
+
+    case class Force[T <: JType](exp: ErasedAst.Expression[JLazy[T]], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[T]
+
+    case class FixpointConstraintSet(cs: List[ErasedAst.Constraint], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[JType]
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala:215]: Result -> JObject

File: main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala [hunk 19]
```diff
@@ -0,0 +1,364 @@
...
+
+    case class FixpointCompose[T <: JType](exp1: ErasedAst.Expression[T], exp2: ErasedAst.Expression[T], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[T]
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala:217]: T -> JObject (everywhere)

File: main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala [hunk 20]
```diff
@@ -0,0 +1,364 @@
...
+
+    case class FixpointSolve(exp: ErasedAst.Expression[JUnknown[Nothing]], stf: Ast.Stratification, tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[JUnknown[Nothing]]
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala:219]: JObject everywhere.

File: main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala [hunk 21]
```diff
@@ -0,0 +1,364 @@
...
+
+    case class FixpointProject(pred: Name.Pred, exp: ErasedAst.Expression[JUnknown[Nothing]], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[JUnknown[Nothing]]
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala:221]: JObject everywhere

File: main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala [hunk 22]
```diff
@@ -0,0 +1,364 @@
...
+
+    case class FixpointEntails(exp1: ErasedAst.Expression[JUnknown[Nothing]], exp2: ErasedAst.Expression[JUnknown[Nothing]], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[PrimInt32]
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala:223]: JUnknown to JObject

File: main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala [hunk 23]
```diff
@@ -0,0 +1,364 @@
...
+
+    case class FixpointFold(pred: Name.Pred, init: ErasedAst.Expression.Var[JUnknown[Nothing]], f: ErasedAst.Expression.Var[JUnknown[Nothing]], constraints: ErasedAst.Expression.Var[JUnknown[Nothing]], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[JType]
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala:225]: Unknown -> JObject

File: main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala [hunk 24]
```diff
@@ -0,0 +1,364 @@
...
+
+    case class HoleError(sym: Symbol.HoleSym, tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[JUnknown[Nothing]]
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala:227]: [T <: JType]

File: main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala [hunk 25]
```diff
@@ -0,0 +1,364 @@
...
+
+    case class MatchError(tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[JUnknown[Nothing]]
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala:229]: [T <: JType]

File: main/src/ca/uwaterloo/flix/language/phase/BytecodeCompiler.scala [hunk 26]
```diff
@@ -0,0 +1,45 @@
+package ca.uwaterloo.flix.language.phase
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/BytecodeCompiler.scala:1]: Move into subfolder `phase/sjvm`.

File: main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala [hunk 27]
```diff
@@ -0,0 +1,364 @@
...
+
+  case class UnboxInt8(exp: ErasedAst.Expression[BoxedInt8], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[PrimInt8]
+  case class UnboxInt16(exp: ErasedAst.Expression[BoxedInt16], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[PrimInt16]
+  case class UnboxInt32(exp: ErasedAst.Expression[BoxedInt32], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[PrimInt32]
+  case class UnboxInt64(exp: ErasedAst.Expression[BoxedInt64], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[PrimInt64]
+  case class UnboxChar(exp: ErasedAst.Expression[BoxedChar], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[PrimChar]
+  case class UnboxFloat32(exp: ErasedAst.Expression[BoxedFloat32], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[PrimFloat32]
+  case class UnboxFloat64(exp: ErasedAst.Expression[BoxedFloat64], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Expression[PrimFloat64]
+
+  sealed trait Expression[+T <: JType] {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala:323]: Move this up just before Object Expression

File: main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala [hunk 28]
```diff
@@ -0,0 +1,364 @@
...
+    sealed trait JObject extends JType
+    sealed trait JUnit extends JType
+
+    sealed trait JRef[T <: JType] extends JType
+    sealed trait JArray[T <: JType] extends JType
+    sealed trait JChan[T <: JType] extends JType
+    sealed trait JLazy[T <: JType] extends JType
+
+    // Indicates an undecided type, this is to be removed
+    sealed trait JUnknown[T <: JType] extends JType
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala:361]: Remove (see above)

File: main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala [hunk 29]
```diff
@@ -0,0 +1,364 @@
...
+
+  object Predicate {
+
+    sealed trait Head extends ErasedAst.Predicate
+
+    object Head {
+
+      case class Atom(pred: Name.Pred, den: Denotation, terms: List[ErasedAst.Term.Head], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Predicate.Head
+
+      case class Union(exp: ErasedAst.Expression[JUnknown[Nothing]], terms: List[ErasedAst.Term.Head], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Predicate.Head
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala:231]: JObject

File: main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala [hunk 30]
```diff
@@ -0,0 +1,364 @@
...
+
+    }
+
+    sealed trait Body extends ErasedAst.Predicate
+
+    object Body {
+
+      case class Atom(pred: Name.Pred, den: Denotation, polarity: Ast.Polarity, terms: List[ErasedAst.Term.Body], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Predicate.Body
+
+      case class Guard(exp: ErasedAst.Expression[JUnknown[Nothing]], terms: List[ErasedAst.Term.Body], loc: SourceLocation) extends ErasedAst.Predicate.Body
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala:241]: JBool

File: main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala [hunk 31]
```diff
@@ -0,0 +1,364 @@
...
+
+    object Head {
+
+      case class QuantVar(sym: Symbol.VarSym, tpe: MonoType, loc: SourceLocation) extends ErasedAst.Term.Head
+
+      case class CapturedVar(sym: Symbol.VarSym, tpe: MonoType, loc: SourceLocation) extends ErasedAst.Term.Head
+
+      case class Lit(sym: Symbol.DefnSym, tpe: MonoType, loc: SourceLocation) extends ErasedAst.Term.Head
+
+      case class App(exp: ErasedAst.Expression[JUnknown[Nothing]], args: List[Symbol.VarSym], tpe: MonoType, loc: SourceLocation) extends ErasedAst.Term.Head
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala:261]: JObject

File: main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala [hunk 32]
```diff
@@ -0,0 +1,364 @@
...
+
+package ca.uwaterloo.flix.language.ast
+
+import ca.uwaterloo.flix.language.ast.Ast.{Denotation, Source}
+import ca.uwaterloo.flix.language.ast.ErasedAst.JType._
+
+import java.lang.reflect.{Constructor, Field, Method}
+
+
+// TODO: replace JType with T <: JType type variable
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ErasedAst.scala:25]: I think this has been done, so the comment can be removed?

**Review by magnus-madsen — COMMENTED** [review]
I added a few comments that should be quick to fix. Then I want to hit the merge button and move onto the next PR.

**Comment by magnus-madsen** [comment]
Ready for merge?

**Comment by JonathanStarup** [comment]
> Ready for merge?

Yes, but you named JBool in a comment, do you want explicit bools or just use PrimInt32 as it is now?

**Comment by magnus-madsen** [comment]
> 
> 
> > Ready for merge?
> 
> Yes, but you named JBool in a comment, do you want explicit bools or just use PrimInt32 as it is now?

My bad. I meant Int32.

**Comment by magnus-madsen** [comment]
Merged. Thanks! Next step: Eraser :)

**Pull request merged**